### PR TITLE
Changes to Test Tree Infrastructure

### PR DESCRIPTION
**Pull request by bryanar**
This work precedes the porting of a large number of tests from an internal test tree. The work includes the following features being added:
- Specifying build instructions as specified by CLRTestKind (BuildOnly, RunOnly, BuildAndRun, and SharedLibrary) - The included documentation/tests.md file has information on the details of the differences here, and how to use them. 
- Test priorities. Given that we are going to have a larger number of tests soon, rather than requiring all tests running all the time, with this set of changes we have the ability to categorize tests by specifying priority values (0+). Priority 0 are to be considered as "Essential" tests, and is also the default value of the test if it is not specified. It is up to the test creator to de-prioritize the test values.
- Ported a small handful of additional GC tests. 
- Dropped the cs_template.csproj files, and instead converted them to individual CS Project files to fit in to the test tree model better.

### CODE REVIEW

**Comment by jasonwilliams200OK** [comment]
@bryanar, your branch <kbd>changes-to-test-tree</kbd> is 35 commits behind the current master. Please rebase:

```powershell
cd coreclr
git remote add dotnet https://github.com/dotnet/coreclr
git checkout changes-to-test-tree
git pull --rebase dotnet master
# fix the merge conflicts
git add .
git rebase --continue
# fix more merge conflicts
git add .
git rebase --continue
# and so on..
# then finally
git push -f
```

**Comment by bryanar** [comment]
Thanks Jason - It is done. Also thanks to @ramarag, I merged my commits.

**Comment by bryanar** [comment]
@gkhanna79 @mmitche @wtgodbe @sergiy-k @russellhadley @yizhang82 @ZhichengZhu @Chrisboh

File: tests/src/CLRTest.Execute.Bash.targets [hunk 1]
```diff
@@ -129,6 +129,24 @@ CLRTestExitCode=$?
       <Output TaskParameter="ParamList" PropertyName="_CLRTestParamList"/>
     </GenerateParamList>
 
+<<<<<<< HEAD
```
Comment (ramarag) [inline tests/src/CLRTest.Execute.Bash.targets]: merge confict have not resolved , you need to do it manually, remove the unnecessary lines

File: tests/src/CLRTest.Execute.Batch.targets [hunk 2]
```diff
@@ -98,12 +101,20 @@ IF NOT "%CLRTestExitCode%"=="%CLRTestExpectedExitCode%" (
       <Output TaskParameter="TargetOutputs" PropertyName="_CLRTestToRunFileFullPath"/>
     </MSBuild>
     
-    <PropertyGroup>
-      <!-- Calculate the thing we're going to run -->
-      <_CLRTestRunFile Condition="'$(_CLRTestNeedsProjectToRun)' != 'True'">"$(AssemblyName).exe"</_CLRTestRunFile>
+    <Message Text="Project depends on $(_CLRTestToRunFileFullPath)." Condition="'$(_CLRTestNeedsProjectToRun)' == 'True'" />
 
+<<<<<<< HEAD
```
Comment (ramarag) [inline tests/src/CLRTest.Execute.Batch.targets]: merge conflict

**Comment by jasonwilliams200OK** [comment]
:+1:
To squash them into a single commit:

```powershell
cd coreclr
git checkout changes-to-test-tree
git rebase -i HEAD~3
# in notepad, replace the word `pick` with `f` in second and third line
# save and close
git push -f
```

CLA signing condition is not met for this PR. Not sure why it is complaining for you, since you are a Collaborator! :)

**Comment by bryanar** [comment]
@jasonwilliams200OK - is it common to continuously squish all commits I make? I.e. should I ALWAYS have my PR 'flat'?

**Comment by jasonwilliams200OK** [comment]
I am not the authority around here, but what I have seen in this kind of feats, where the subsequent commits are not logically segregated (but merely about `addressing feedback`, `merging master`) repo owners suggest to squash / fixup those extra commits (which will eventually make the source tree noisy and if you have some commit in the middle failing the CI build, it will make it harder for people to `git bisect` etc. to spot the regression bugs in future).

So if you have opened a PR with a single commit and you wana just make some changes, you can simply:

```powershell
cd coreclr
# notepad blah.foo
# or
# devenv blah.foo
#
# make changes
# then
git commit --amend
# close editor
git push -f
```

to continuously amend the same "last" commit.

If you have to multiple commits in a PR then amending the commits other than the last one normally require interactive rebase `git rebase -i HEAD~<count n backwards from the tip of your branch>` etc. I usually prefer this approach. There are also other methods to do it as with git, there is always more than one way to skin a cat. :smile_cat:

**Comment by bryanar** [comment]
Very useful advisement, thanks Jason!

**Comment by ellismg** [comment]
FYI @bryanar, Microsoft employees still need to sign the CLA due to the Microsoft/.NET Foundation split.

**Comment by jasonwilliams200OK** [comment]
@dotnet-bot, test this please.

**Pull request merged**